node-manager/src/dlc.rs

```diff
@@ -0,0 +1,275 @@
+use crate::localstorage::MutinyBrowserStorage;
+use bdk::blockchain::EsploraBlockchain;
+use bdk::database::Database;
+use bdk::keys::ExtendedKey;
+use bdk::wallet::AddressIndex;
+use bdk::{SignOptions, Wallet};
+use bdk_macros::maybe_await;
+use bip39::Mnemonic;
+use bitcoin::psbt::PartiallySignedTransaction;
+use bitcoin::util::address::Address;
+use bitcoin::util::bip32::{ChildNumber, DerivationPath, ExtendedPrivKey};
+use bitcoin::{Block, Network, Script, Transaction, TxIn, Witness};
+use dlc_manager::error::Error as ManagerError;
+use dlc_manager::error::Error::WalletError;
+use dlc_manager::{Blockchain, Utxo};
+use log::error;
+use secp256k1::Secp256k1;
+use std::str::FromStr;
+use std::sync::Arc;
+use wasm_bindgen_futures::spawn_local;
+
+#[derive(Debug)]
+pub struct MutinyDLCWallet {
+    wallet: Arc<Wallet<MutinyBrowserStorage>>,
+    blockchain: Arc<EsploraBlockchain>,
+    storage: Arc<MutinyBrowserStorage>,
+    mnemonic: Mnemonic,
+}
+
+fn bdk_err_to_manager_err(e: bdk::Error) -> ManagerError {
+    WalletError(Box::new(e))
+}
+
+impl MutinyDLCWallet {
+    fn derive_secret_key(&self, index: u32) -> secp256k1::SecretKey {
+        let entropy = self.mnemonic.to_entropy();
+        let network = self.wallet.network();
+        let xprivkey = ExtendedPrivKey::new_master(network, &entropy).unwrap();
+        let xkey: ExtendedKey = ExtendedKey::from(xprivkey);
+        let master_xprv = xkey.into_xprv(network).unwrap();
+        let coin_type = match network {
+            Network::Bitcoin => 0,
+            Network::Testnet => 1,
+            Network::Signet => 1,
+            Network::Regtest => 1,
+        };
+
+        let base_path = DerivationPath::from_str("m/585'").unwrap();
+
+        let derivation_path = base_path.extend([
+            ChildNumber::from_hardened_idx(coin_type).unwrap(),
+            ChildNumber::from_hardened_idx(index).unwrap(),
+        ]);
+
+        let k = master_xprv
+            .derive_priv(&Secp256k1::new(), &derivation_path)
+            .unwrap();
+
+        k.private_key
+    }
+}
+
+impl dlc_manager::Signer for MutinyDLCWallet {
+    fn sign_tx_input(
+        &self,
+        tx: &mut bitcoin::blockdata::transaction::Transaction,
+        input_index: usize,
+        tx_out: &bitcoin::blockdata::transaction::TxOut,
+        redeem_script: Option<bitcoin::blockdata::script::Script>,
+    ) -> Result<(), ManagerError> {
```

Comment:
Now that we've got multiple "wallets", maybe putting this inside some generic BDK wallet wrapper could be useful. Maybe not now but having multiple "wallets" and "chains" can get rather confusing and probably some leakage in terms of reusing logic. I don't think I see anything DLC specific here.

Reply:
yeah this is just kinda a hack to sign a specific input, but I agree that things could probably be cleaned up. That is kinda what I want to propose for the spiral grant lol